README.md
# mongodb-local-data-api

Run your own version of the [MongoDB Atlas Data API](https://docs.atlas.mongodb.com/api/data-api/) for local development and testing.

This is a simple HTTP server that *simulates* the [Data API endpoints](https://docs.atlas.mongodb.com/api/data-api-resources/). It was designed for exact parity with the Atlas version, it was not designed for use in production.

> **v1 Update!** MongoDB's "Data API" is out of beta, so this library is now v1 as well! 🎉

## Example

Suppose you have a MongoDB instance running locally, following [these Docker instructions](https://www.mongodb.com/compatibility/docker), and you want to also run a local version of the Data API for development:

```bash
mongodb-local-data-api --username=AzureDiamond \
  --password=hunter2 \
  --mongodbPort=27017 \
  --apiPort=3001 \
  --key=battery-horse-staple
```

That starts this mock-Data-API with the [base URL](https://www.mongodb.com/docs/atlas/api/data-api-resources/#base-url) as `http://localhost:3001`, so e.g. to [find a single document](https://www.mongodb.com/docs/atlas/api/data-api-resources/#find-a-single-document) (which is the `/action/findOne` path) with `cURL` you would do:

```bash
curl --request POST \
  'http://localhost:3000/action/findOne' \
  --header 'Content-Type: application/json' \
  --header 'api-key: battery-horse-staple' \
  --data-raw '{
      "dataSource": "Cluster0",
      "database": "todo",
      "collection": "tasks",
      "filter": {
        "text": "Do the dishes"
      }
  }'
```

You can make any request to this API that is in the [official documentation](https://www.mongodb.com/docs/atlas/api/data-api-resources/#run-an-aggregation-pipeline).

You can interact with this library programmatically, if you'd rather:

```js
import { setup } from 'mongodb-local-data-api'
const database = setup({
	// the URL to access the locally-running MongoDB instance
	url: 'mongodb://AzureDiamond:hunter2@localhost:27017',
})
// correlates to the MongoDB action, e.g. find, findOne, aggregate, ...
const action = 'insertOne'
// these are the parameters as you would send them to the Data API
const parameters = {
	dataSource: 'Cluster0',
	database: 'todo',
	collection: 'tasks',
	document: {
		status: 'open',
		text: 'Do the dishes'
	}
}
const { status, body } = await database(action, parameters)
```

## Options

These are the options you can pass to the `setup` function:

- `url: String` - The fully qualified URL to the MongoDB instance. This would look like `mongodb://AzureDiamond:hunter2@localhost:27017` or similar. If you set this property, the remaining MongoDB-related options will be ignored.
- `username: String` - The username to access the MongoDB instance.
- `password: String` - The password to access the MongoDB instance.
- `dbDomain: String` - The domain of the MongoDB instance. (Default: `localhost`)
- `dbPort: Integer` - The port of the MongoDB instance. (Default: `27017`)
- `retryCount: Integer` - By default any interrupted connections to MongoDB will be retried indefinitely. Set to `0` to disable, or set to a limit. (Default: `Infinity`)
- `verbose: Boolean` - Whether to print out additional information to the console. (Default: `false`)

The options you can pass to the CLI are the same as above, but include these options:

- `apiPort: Integer` - The port used to access this Data API instance. (Default: `3007`)
- `key: String` - Authentication keys to look for on the `API-Key` request header, to access the local API. Set this option multiple times for multiple keys. Default: no `API-Key` required.

For the CLI tool, all options ***except `key`*** can also be set using the environment variable version instead, which is simply the prefix `MONGODBLOCAL_` followed by the option name, e.g. `MONGODBLOCAL_URL`. The environment variable is case-insensitive, so e.g. `MONGODBLOCAL_url` or `MongoDBLocal_url` would also work.

## Example

Here's an example of how I've used this library to run a local environment, with MongoDB in a Docker container, and an `npm` run script to coordinate.

I set up a container with services, using `docker compose`. (In most of my applications I had additional other services in the same container.)

```yaml
# $REPO/docker/docker-compose.yaml
version: '3.8'
services:
  mongodb-local:
    image: mongo:5.0
    ports:
      - "27017:27017"
    # this is how you get MongoDB to stop spewing out so many logs
    command: mongod --quiet --logpath /dev/null
    environment:
      - MONGO_INITDB_ROOT_USERNAME=AzureDiamond
      - MONGO_INITDB_ROOT_PASSWORD=hunter2
    volumes: # this persists MongoDB data between restarts
      - "./mongodb:/data/db"
```

Then in my repo root, in the `package.json` file I have these relevant parts:

```json
{
  "scripts": {
    "local": "run-p local:*",
    "local:docker": "cd deploy && docker compose up",
    "local:mongodb-data-api": "mongodb-local-data-api --username=AzureDiamond --password=hunter2 --apiKey=battery-horse-staple --apiPort=3007"
  },
  "devDependencies": {
    "mongodb-local-data-api": "^1.0.0",
    "npm-run-all": "^4.0.0"
  }
}
```

Then when you do `npm run local` it launches the Docker container (using docker-compose) and the local Data API, in parallel.

## Related

You might also be interested in [`@saibotsivad/mongodb`](https://github.com/saibotsivad/mongodb) which is a thin wrapper to interact with a MongoDB Atlas Data API instance:

```js
import { mongodb } from '@saibotsivad/mongodb'

const db = initializeMongodb({
	apiUrl: process.env.MONGODB_API_URL,
	apiKey: process.env.MONGODB_API_KEY,
	dataSource: process.env.MONGODB_DATA_SOURCE,
	database: process.env.MONGODB_DATABASE_NAME,
	collection: process.env.MONGODB_COLLECTION_NAME,
})

const { documents } = await db.findOne({
	filter: {
		text: 'Do the dishes',
	},
})
```

## License

Published and released under the [Very Open License](http://veryopenlicense.com).

If you need a commercial license, [contact me here](https://davistobias.com/license?software=mongodb-local-data-api).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Change categories are:

* `Added` for new features.
* `Changed` for changes in existing functionality.
* `Deprecated` for once-stable features removed in upcoming releases.
* `Removed` for deprecated features removed in this release.
* `Fixed` for any bug fixes.
* `Security` to invite users to upgrade in case of vulnerabilities.

## Unreleased
### Added
### Changed
### Deprecated
### Fixed
### Removed
### Security

## [1.0.1](https://github.com/saibotsivad/mongodb-local-data-api/compare/v0.0.11...v1.0.1) - 2022-10-06
### Added
- Documentation to show plain `cURL` interaction with the API, and make the demo work better.
### Changed
- The endpoints that previously returned `Content-Type: text/plain` from the **beta** version of the Data API have been updated to correctly return `application/json` so this library does as well. Closes #1
### Fixed
- The endpoints now enforce `application/json` a bit more strictly on outputs that return it.

## [0.0.10-0.0.11](https://github.com/saibotsivad/mongodb-local-data-api/compare/v0.0.9...v0.0.11) - 2022-09-08
### Fixed
- The `aggregate` function was being called incorrectly.

## [0.0.9](https://github.com/saibotsivad/mongodb-local-data-api/compare/v0.0.8...v0.0.9) - 2022-07-13
### Fixed
- The `deleteMany` was being called incorrectly.

## [0.0.7-0.0.8](https://github.com/saibotsivad/mongodb-local-data-api/compare/v0.0.6...v0.0.8) - 2022-05-25
### Added
- If scanned results is over 1000, will log an error. (This more closely matches the behaviour of MongoDB Atlas.)
### Changed
- Log output with and without the `--verbose` flag is cleaner.

## [0.0.6](https://github.com/saibotsivad/mongodb-local-data-api/compare/v0.0.5...v0.0.6) - 2022-02-10
### Fixed
- The `findOne` method being called incorrectly. 🤦‍♂️ I hadn't noticed until now because I so rarely use that specific method.

## [0.0.4-0.0.5](https://github.com/saibotsivad/mongodb-local-data-api/compare/v0.0.3...v0.0.5) - 2022-01-28
### Added
- If the connection to MongoDB gets dropped, it'll try to reconnect. You can control the number of times using the `retryCount` option.
### Fixed
- Added the `verbose` option to the docs.
- Corrected the examples for `sade`.

## [0.0.2-0.0.3](https://github.com/saibotsivad/mongodb-local-data-api/compare/v0.0.1...v0.0.3) - 2022-01-27
### Fixed
- The `insertOne` and `find` operations were incorrect.

## [0.0.1](https://github.com/saibotsivad/mongodb-local-data-api/compare/v0.0.0...v0.0.1)
### Fixed
- Docs for options were incomplete.

## [0.0.0](https://github.com/saibotsivad/mongodb-local-data-api/tree/v0.0.0) - 2022-01-26
### Added
- Created the base project from [saibotsivad/init](https://github.com/saibotsivad/init).
- First pass at making it an accurate copy of the existing Data API, warts and all.


src/index.js
import { MongoClient } from 'mongodb'

const supportedActions = [
	'aggregate',
	'deleteOne',
	'deleteMany',
	'find',
	'findOne',
	'insertOne',
	'insertMany',
	'replaceOne',
	'updateOne',
	'updateMany',
]

const actionValidation = {
	insertOne: ({ document }) => {
		if (!document) return {
			status: 400,
			body: "Failed to insert document: must provide a 'document' field",
		}
	},
}

const specialAction = {
	aggregate: (coll, params) => coll.aggregate(params.pipeline || []).toArray().then(documents => ({ documents })),
	deleteMany: (coll, params) => {
		const { filter, ...remaining } = params
		return coll.deleteMany(filter || {}, remaining || {})
	},
	find: (coll, params) => {
		const { filter, ...remaining } = params
		return coll.find(filter, remaining || {}).toArray().then(documents => ({ documents }))
	},
	findOne: (coll, params) => {
		const { filter, ...remaining } = params
		return coll.findOne(filter, remaining).then(document => ({ document }))
	},
	insertMany: async (coll, params) => {
		const { documents, ...remaining } = params
		const { insertedIds } = await coll.insertMany(documents, remaining)
		return { insertedIds: Object.values(insertedIds).map(o => o.toString()) }
	},
	insertOne: async (coll, params) => coll.insertOne(params.document),
	replaceOne: async (coll, params) => {
		const { filter, replacement, ...remaining } = params
		const { matchedCount, modifiedCount, upsertedId } = await coll.replaceOne(filter, replacement, remaining)
		const out = { matchedCount, modifiedCount }
		if (upsertedId) out.upsertedId = upsertedId
		return out
	},
	updateOne: async (coll, params) => {
		const { filter, update, ...remaining } = params
		const { matchedCount, modifiedCount, upsertedId } = await coll.updateOne(filter, update, remaining)
		const out = { matchedCount, modifiedCount }
		if (upsertedId) out.upsertedId = upsertedId
		return out
	},
	updateMany: async (coll, params) => {
		const { filter, update, ...remaining } = params
		const { matchedCount, modifiedCount, upsertedId } = await coll.updateMany(filter, update, remaining)
		const out = { matchedCount, modifiedCount }
		if (upsertedId) out.upsertedId = upsertedId
		return out
	},
}

const tidyResults = (actionName, body) => {
	if (body && body.acknowledged) delete body.acknowledged
	if (body && body.insertedCount) delete body.insertedCount
	return body
}

export const setup = ({ url, verbose, retryCount }) => {
	const databaseToConnection = {}
	return async (actionName, { dataSource, database, collection, ...params }) => {
		if (dataSource && verbose) console.warn('The property `dataSource` is currently ignored for local databases.')
		if (!supportedActions.includes(actionName)) return { status: 404, body: '' }

		const validationError = actionValidation[actionName] && actionValidation[actionName](params)
		if (validationError) return validationError

		let running
		const run = async () => {
			if (!databaseToConnection[database]) {
				console.log(`Attempting to connect to: ${database}`)
				const client = new MongoClient(url)
				await client.connect()
				databaseToConnection[database] = client.db(database)
			}
			const coll = databaseToConnection[database].collection(collection)
			const results = await (specialAction[actionName] ? specialAction[actionName](coll, params) : coll[actionName](params))
			running = true
			return {
				count: results?.documents?.length,
				status: actionName.startsWith('insert') ? 201 : 200,
				body: tidyResults(actionName, results),
			}
		}

		try {
			return await run()
		} catch (firstRunError) {
			if (firstRunError.message.includes('ECONNREFUSED')) {
				console.log('Connection to MongoDB was interrupted, trying again...')
				let retries = 1
				while (!running && (retries < retryCount || retryCount === undefined)) {
					try {
						return await run()
					} catch (retryError) {
						if (!retryError.message.includes('ECONNREFUSED')) throw retryError
						console.log(`Reconnect retry ${++retries} of ${retryCount === undefined ? '∞' : retryCount}...`)
					}
				}
			}
			console.log('Error handling request:', firstRunError.message)
			return {
				status: 400,
				body: firstRunError.message,
			}
		}
	}
}


demo/insert-item.js
import { post } from 'httpie'
import { mongodb } from '@saibotsivad/mongodb'

import { setTimeout as delay } from 'node:timers/promises'
await delay(1000) // artificial wait time until the local API starts

const API_URL = 'http://localhost:9001'
const API_KEY = 'abc123'

/*

You can call the locally running API manually, with a plain ol' HTTP request:

 */
post(API_URL + '/action/insertOne', {
	headers: {
		'Content-Type': 'application/json',
		'Api-Key': API_KEY,
	},
	body: JSON.stringify({
		dataSource: 'Cluster0',
		database: 'office',
		collection: 'people',
		document: { name: 'John Jacob Jingleheimerschmidt' },
	}),
})
	.then(response => {
		console.log('done', {
			status: response.statusCode,
			type: response.headers['content-type'],
			data: response.data,
		})
	})
	.catch(error => {
		console.log('bad request', {
			status: error.statusCode,
			type: error.headers['content-type'],
			data: error.data,
		})
	})

/*
 Ignore this, it's a small shim to make `httpie` behave closer to the `fetch` specs.
 */
const postFetchShim = response => ({
	status: response.statusCode,
	json: async () => response.data,
	text: async () => response.data,
})
const fetch = async (url, parameters) => post(url, parameters).then(postFetchShim, postFetchShim)
/*

You can also use handy dandy tooling to make it easier:

 */
const db = mongodb({
	apiUrl: API_URL,
	apiKey: API_KEY,
	dataSource: 'Cluster0',
	database: 'office',
	collection: 'people',
	fetch,
})
db
	.insertOne({
		document: { name: 'John Jacob Jingleheimerschmidt' },
	})
	.then(out => {
		console.log('huzza!', out)
	})
	.catch(error => {
		console.log('oh no!', error)
	})


bin.js
#!/usr/bin/env node

import { readFileSync } from 'node:fs'
import sade from 'sade'
import Koa from 'koa'
import koaRouter from 'koa-router'
import koaBody from 'koa-body'
import { setup } from './src/index.js'

const optionNames = [
	'apiPort',
	'url',
	'username',
	'password',
	'dbDomain',
	'dbPort',
]

const mixWithEnv = callback => options => {
	const envOptions = {}
	for (const envName in process.env) {
		const tidyName = envName.toUpperCase()
		if (tidyName.startsWith('MONGODBLOCAL_')) envOptions[tidyName] = process.env[envName]
	}
	for (const opt of optionNames) {
		if (options[opt] === undefined && envOptions[`MONGODBLOCAL_${opt.toUpperCase()}`]) {
			options[opt] = envOptions[`MONGODBLOCAL_${opt.toUpperCase()}`]
		}
	}
	if (options.apiPort) options.apiPort = parseInt(options.apiPort, 10)
	if (options.dbPort) options.dbPort = parseInt(options.dbPort, 10)
	if (options.retryCount) options.retryCount = parseInt(options.retryCount, 10)
	callback(options)
}

sade('mongodb-local-data-api', true)
	.version(JSON.parse(readFileSync('./package.json', 'utf8')).version)
	.describe('Start an instance of a Data API server. Use environment variables instead by prefixing `MONGODBLOCAL_` to the option, e.g. `MONGODBLOCAL_PORT` for `port`.')
	.option('--apiPort', 'Change the port this Data API uses. (Default: 3007)')
	.option('--key', 'The API key to require on all requests. The default is no authentication required. Set the flag multiple times for multiple keys.')
	.option('--url', 'The complete MongoDB URL, e.g. `mongodb://AzureDiamond:hunter2@localhost:27017`. If set, this will override all other related properties.')
	.option('--username', 'The MongoDB username, e.g. `AzureDiamond`.')
	.option('--password', 'The MongoDB password, e.g. `hunter2`.')
	.option('--dbDomain', 'Change the port used to connect to MongoDB. (Default: localhost)')
	.option('--dbPort', 'Change the port used to connect to MongoDB. (Default: 27017)')
	.option('--retryCount', 'By default interrupted MongoDB connections will retry indefinitely. Set to `0` to disable, or an integer for a retry limit. (Default: Infinity)')
	.option('--verbose', 'If this flag is set, all request and response bodies will also be logged.')
	.example('--username=AzureDiamond --password=hunter2 # normal')
	.example('--url=mongodb://user:pass@localhost:27017/?serverSelectionTimeoutMS=5000 # connection options')
	.example('--key=abc123 --key=def456 # multiple keys')
	.action(mixWithEnv(({ apiPort, url, username, password, dbDomain, dbPort, key: keys, verbose, retryCount }) => {
		if (!Array.isArray(keys)) keys = keys ? [ keys ] : []
		if (!apiPort) apiPort = 3007
		if (!dbPort) dbPort = 27017
		if (!dbDomain) dbDomain = 'localhost'
		if (!url) url = `mongodb://${username}:${password}@${dbDomain}:${dbPort}`
		const database = setup({ url, verbose, retryCount })
		const app = new Koa()
		const router = koaRouter()
		let requestCounter = 0
		router
			.post('/action/:action', koaBody(), async context => {
				const action = context.request.params.action
				const requestId = (++requestCounter).toString().padStart(4, '0')
				if (verbose) console.log(requestId, new Date(), action, context.request.body)
				if (keys.length && !keys.includes(context.request.headers['api-key'])) {
					console.log(requestId, new Date(), 'Provided API key was not authorized:', context.request.headers['api-key'])
					context.status = 401
					context.headers['content-type'] = 'application/json'
					context.body = {
						error: 'invalid session: error finding user for endpoint',
						error_code: 'InvalidSession',
						link: 'https://realm.mongodb.com/groups/611be2cf39b99f2a25556d10/apps/61df4d50cb330d69cb2b7822/logs?co_id=61f1c4aaa37cb8585dbdd69c',
					}
				} else {
					await database(action, context.request.body)
						.then(({ status, body, count }) => {
							context.status = status
							context.body = body
							if (typeof body !== 'string') context.headers['content-type'] = 'application/json'
							let message = `action: ${action}; status: ${status}`
							if (count) message += `; documents: ${count}`
							console.log(requestId, new Date(), message)
							if (count > 1000) console.error(requestId, new Date(), 'Error: Returned Objects has gone above 1000.', context.request.body)
							if (verbose) console.log(body)
						})
						.catch(error => {
							console.error('The database request failed for an unexpected reason:', error)
							process.exit(1)
						})
				}
			})
		app.use(router.routes())
		app.listen(apiPort, () => {
			console.log(new Date(), `Data API running on port ${apiPort}`)
		})
	}))
	.parse(process.argv)
